### ISSUE TITLE
`NativeFunction` does not really support variadic arguments

### ISSUE THREAD

**Issue by gebing** [frida/frida-gum#366]
When i modified some unit tests of frida-gum, i found `NativeFunction` did not really support variadic arguments. Below is the original test code for `variadic_native_function_can_be_invoked` and it can work fine because the calling arguments exactly equals to arguments count of function declaration:

```
  COMPILE_AND_LOAD_SCRIPT (
      "var sum = new NativeFunction(" GUM_PTR_CONST ", "
          "'int', ['pointer', 'pointer', '...', 'float']);"
      "send(sum(ptr(3), ptr(4), 42.0));",
      gum_add_pointers_and_float_variadic);
  EXPECT_SEND_MESSAGE_WITH ("49");
  EXPECT_NO_MESSAGES ();
```

But if i add more arguments when calling the variadic function, it will failed with error of `bad argument count`. Below is the code i modified:

```
  COMPILE_AND_LOAD_SCRIPT (
      "var sum = new NativeFunction(" GUM_PTR_CONST ", "
          "'int', ['pointer', 'pointer', '...', 'float']);"
      "send(sum(ptr(3), ptr(4), 42.0, 100.0, 200.0));",
      gum_add_pointers_and_float_variadic);
  EXPECT_SEND_MESSAGE_WITH ("349");
  EXPECT_NO_MESSAGES ();
```

I found it is because the `gum_duk_native_function_invoke` and `gum_v8_native_function_invoke` do not process the variadic calling arguments which passed to function. I will make a pull request for fixing this issue.

**Comment by oleavr** [comment]
This is unfortunately the way libffi works – you have to create one `NativeFunction` per unique signature. But if you see a way we can improve our abstraction to maybe hide that detail, without a performance regression, that would be great.

**Comment by gebing** [comment]
I looked at the source code of [pyobjc](https://bitbucket.org/ronaldoussoren/pyobjc). It's solution is that if it is a variadic function, it will postpone the calling of `ffi_prep_cif` to the time when the function is invoked. At that time, we can know the exact number and type of arguments. 

What do you think about this solution? If you agree, i can make a pull request based on it.

@oleavr

**Comment by oleavr** [comment]
Yes, a CIF needs to be created per unique signature. So what we could support is that you just put `'...'` as the last argument, and then whenever you invoke it we can compute an ID based on the signature and check if we've already got a CIF pre-generated, and generate one if not. (E.g. using a `GHashTable` to keep track of the different pre-generated CIFs.)

**Comment by gebing** [comment]
I made a pull request for implement variadic function for `NativeFunction`. 

The variadic `NativeFunction` can be declared such as `var func = new NativeFunction(ptr, 'int', ['pointer', '...', 'pointer', 'float'])`, which means the variadic arguments type can be `'pointer', 'float', 'pointer', 'float',...` repeatedly. And the invocation of variadic function can be with any number of arguments which not less than the number of fixed arguments, such as `func(ptr(1))` or `func(ptr(1), ptr(2))`

@oleavr 
I do not use the GHashTable to keep the calling signature, because i looked into source of function `ffi_prep_cif`, and found there is only some value assignment, so i think it will cause only little performance regression. I also made some optimization for only call `ffi_prep_cif` function when the provided arguments count is larger than declaration's arguments count.
